src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala

```diff
@@ -1,32 +1,76 @@
+import edg.wir.ProtoUtil.BlockProtoToSeqMap
 import edg_ide.swing.ElkNodePainter
+import edgir.elem.elem.{BlockLike, HierarchyBlock}
 import org.eclipse.elk.graph.ElkNode
+import edg.wir.{DesignPath}
+import edg_ide.edgir_graph.HierarchyGraphElk
 
 import java.awt.Color
-import java.io.FileOutputStream
+import java.io.{FileNotFoundException, FileOutputStream, IOException}
 
 
 object PDFGeneratorUtil{
 
-  def generate(rootNode: ElkNode, fileName: String): Unit = {
-    // TODO: Set a fixed document size and scale graphics accordingly?
-    val width = rootNode.getWidth.toFloat + 2 * ElkNodePainter.margin.toFloat
-    val height = rootNode.getHeight.toFloat + 2 * ElkNodePainter.margin.toFloat
-    val document = new Document(new Rectangle(width, height))
+  private def generatePageSize(node: ElkNode): (Float, Float) = {
+    val width = node.getWidth.toFloat + 2 * ElkNodePainter.margin.toFloat
+    val height = node.getHeight.toFloat + 2 * ElkNodePainter.margin.toFloat
+    (width, height)
+  }
+
+  def generate(content: HierarchyBlock, fileName: String): Unit = {
+    try {
+      val document = new Document()
+      val writer = PdfWriter.getInstance(document, new FileOutputStream(fileName))
+      val footer = new HeaderFooter(true)
+      footer.setBorder(Rectangle.NO_BORDER)
+      footer.setAlignment(Element.ALIGN_RIGHT)
+      document.setFooter(footer)
+
+      def printNode(node: ElkNode): Unit = {
+        val (width, height) = generatePageSize(node)
+        document.setPageSize(new Rectangle(width, height))
+
+        /*
+        Metadata for the Footer does not align the page number correctly if
+        document.open() is called before document.setPageSize() was executed
+         */
+        if (!document.isOpen) {
+          document.open()
+        } else {
+          document.newPage
+        }
+
+        val cb = writer.getDirectContent
+        val graphics = cb.createGraphics(width, height)
+        val painter = new ElkNodePainter(node)
+        painter.paintComponent(graphics, Color.white)
+        graphics.dispose()
+      }
 
-    // TODO: make a try..catch for FileOutputStream
-    val writer = PdfWriter.getInstance(document, new FileOutputStream(fileName))
-    document.open()
-    val cb = writer.getDirectContent
-    val graphics = cb.createGraphics(width, height)
+      def printNextHierarchyLevel(block: HierarchyBlock, path: DesignPath = DesignPath()): Unit = {
+        val node = HierarchyGraphElk.HBlockToElkNode(block, path)
+        printNode(node)
 
-    val painter = new ElkNodePainter(rootNode)
-    painter.paintComponent(graphics, Color.white)
+        block.blocks.asPairs.map {
+          case (name, subblock) => (name, subblock.`type`)
+        }
+          .collect {
+            case (name, BlockLike.Type.Hierarchy(subblock)) if subblock.blocks.nonEmpty => (path + name, subblock)
+          }.toMap.foreach {
+          case (path, subblock) => printNextHierarchyLevel(subblock, path)
+        }
```

Comment:
```suggestion
        block.blocks.asPairs.map {
          case (name, subblock) => (name, subblock.`type`)
        }.collect {
          case (name, BlockLike.Type.Hierarchy(subblock)) if subblock.blocks.nonEmpty => (path + name, subblock)
        }.toMap.foreach {
          case (path, subblock) => printNextHierarchyLevel(subblock, path)
        }
```

Minor style suggestion, this helps keep similar functionality at similar levels of indentation.
Also, is the `toMap` necessary?